protocol/consistencychecks.go

```diff
@@ -42,222 +38,243 @@ type ConsistencyChecks struct {
 	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
-		return invalidProof, ErrorBadVRFProof
+		return ErrorBadVRFProof
 	}
 
-	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { // proof of inclusion */
-		// verify name-to-key binding
+	if key == nil {
 		// key is nil when the user does lookup for the first time
-		if key != nil && !ap.VerifyBinding(key) {
-			return invalidProof, ErrorBadBinding
-		}
-		// verify commitment
-		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
-			return invalidProof, ErrorBadCommitment
-		}
-		proofType = proofOfInclusion
+		key = ap.Leaf.Value
 	}
 
 	// verify auth path
-	if !ap.Verify(str.TreeHash) {
-		return invalidProof, ErrorBadAuthPath
+	if !ap.Verify([]byte(uname), key, str.TreeHash) {
+		return ErrorBadAuthPath
 	}
 
-	return proofType, nil
+	return nil
 }
 
-// verifySTR checks the consistency between 2 snapshots.
-// This uses the pinning signing key in cc
+// verifySTRConsistency checks the consistency between 2 snapshots.
+// This should be called if the request is either registration or
+// monitoring. It uses the pinning signing key in cc
 // to verify the STR's signature and should not verify
-// the hash chain using the current STR stored in cc.
-func (cc *ConsistencyChecks) verifySTR(savedSTR, str *m.SignedTreeRoot) error {
+// the hash chain using the STR stored in cc.
+func (cc *ConsistencyChecks) verifySTRConsistency(savedSTR, str *m.SignedTreeRoot) error {
 	// verify STR's signature
 	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
 		return ErrorBadSignature
 	}
-
-	// verify hash chain
-	// TODO: clearly verify which is the client's actual expectation.
-	// See #81
-	if (str.Epoch == savedSTR.Epoch && bytes.Equal(savedSTR.Signature, str.Signature)) ||
-		(str.Epoch == savedSTR.Epoch+1 && str.VerifyHashChain(savedSTR.Signature)) {
+	if str.VerifyHashChain(savedSTR) {
 		return nil
 	}
 
 	// TODO: verify the directory's policies as well. See #115
 	return ErrorBadSTR
 }
 
-// verifyFulfilledPromise verifies issued TBs was inserted
+// verifySTR checks whether the received STR is the same with
+// the SavedSTR using reflect.DeepEqual().
+// This should be called if the request is lookup, since the SavedSTR
+// should only be updated in the beginning of each epoch by registration
+// or monitoring requests.
+func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
+	if reflect.DeepEqual(cc.SavedSTR, str) {
+		return nil
+	}
+	return ErrorBadSTR
+}
+
+// verifyFulfilledPromise verifies issued TBs were inserted
 // in the directory as promised.
 func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
 	str *m.SignedTreeRoot,
 	ap *m.AuthenticationPath) error {
-	if !cc.useTBs {
-		return nil
-	}
-
-	if cc.SavedSTR.Epoch+1 != str.Epoch {
-		return nil
-	}
-
-	tb := cc.TBs[uname]
-	if tb != nil {
-		if !tb.Verify(ap) {
+	if tb, ok := cc.TBs[uname]; ok {
+		if str.Epoch != cc.oldSTR.Epoch+1 {
+			return ErrorBadPromise
```

Comment:
> I thought that we could use KeyLookup for monitoring until we support it.

Oh, I see. I think this would actually be fine, and just add a FIXME. Let me look at the new changes, though.